src/apple/macos/ffi.rs

```diff
@@ -154,3 +177,133 @@ mod io_service {
 mod io_service {}
 
 pub use io_service::*;
+
+#[repr(C)]
+pub struct __IOHIDServiceClient(libc::c_void);
+
+pub type IOHIDServiceClientRef = *const __IOHIDServiceClient;
+
+#[repr(C)]
+pub struct __IOHIDEvent(libc::c_void);
+
+pub type IOHIDEventRef = *const __IOHIDEvent;
+
+pub const kIOHIDEventTypeTemperature: i64 = 15;
+
+pub const kIOHIDEventTypePower: i64 = 15;
+
+pub fn IOHIDEventFieldBase(event_type: i64) -> i64 {
+    event_type << 16
+}
+
+#[cfg(not(feature = "apple-sandbox"))]
+extern "C" {
+    pub fn IOHIDEventSystemClientCreate(allocator: CFAllocatorRef) -> *mut libc::c_void;
+    
+    pub fn IOHIDEventSystemClientSetMatching(client: *mut libc::c_void, matches: CFDictionaryRef) -> i32;
+
+    pub fn IOHIDEventSystemClientCopyServices(client: *mut libc::c_void) -> CFArrayRef;
+
+    pub fn IOHIDServiceClientCopyProperty(service: IOHIDServiceClientRef, key: CFStringRef) -> CFStringRef;
+
+    pub fn IOHIDServiceClientCopyEvent(service: IOHIDServiceClientRef, v0: i64, v1: i32, v2: i64) -> IOHIDEventRef;
+
+    pub fn IOHIDEventGetFloatValue(event: IOHIDEventRef, field: i64) -> f64;
+}
+
+pub type CFMutableArrayRef = *mut __CFArray;
+
+extern "C" {
+    pub fn CFArrayCreateMutable(allocator: CFAllocatorRef, capacity: CFIndex, callbacks: *const CFArrayCallBacks) -> CFMutableArrayRef;
+
+    pub fn CFArrayAppendValue(the_array: CFMutableArrayRef, value: *const libc::c_void);
+}
+
+pub(crate) const HID_DEVICE_PROPERTY_VENDOR_ID: &str = "VendorId";
+pub(crate) const HID_DEVICE_PROPERTY_PRODUCT_ID: &str = "ProductID";
+pub(crate) const HID_DEVICE_PROPERTY_PRODUCT: &str = "Product";
+pub(crate) const HID_DEVICE_PROPERTY_PRIMARY_USAGE: &str = "PrimaryUsage";
+pub(crate) const HID_DEVICE_PROPERTY_PRIMARY_USAGE_PAGE: &str = "PrimaryUsagePage";
+pub(crate) const HID_DEVICE_PROPERTY_MAX_INPUT_REPORT_SIZE: &str = "MaxInputReportSize";
+pub(crate) const HID_DEVICE_PROPERTY_MAX_OUTPUT_REPORT_SIZE: &str = "MaxOutputReportSize";
+pub(crate) const HID_DEVICE_PROPERTY_REPORT_ID: &str = "ReportID";
+
+pub(crate) const kHIDPage_AppleVendor: i32 = 0xff00;
+pub(crate) const kHIDUsage_AppleVendor_TemperatureSensor: i32 = 0x0005;
+
+pub(crate) fn matching(page: i32, usage: i32) -> CFDictionaryRef {
+    unsafe {
+        let primary_usage_page = CString::new(HID_DEVICE_PROPERTY_PRIMARY_USAGE_PAGE).unwrap();
```

Comment:
Instead of going through a `CString`, you could change the constant to look like this:

```rust
const HID_DEVICE_PROPERTY_PRIMARY_USAGE_PAGE: &[u8] = b"PrimaryUsage\0";
```

And then use it like this:

```rust
HID_DEVICE_PROPERTY_PRIMARY_USAGE_PAGE.as_ptr()
```

The `\0` is already at the end so no need for a `CString` (which would perform an extra allocation).